File: python/15_dynamic_programming.py
# Dynamic programming: top-down memoization

def fn(arr):
    def dp(STATE):
        if "BASE_CASE":
            return 0
        
        if STATE in memo:
            return memo[STATE]
        
        ans = "RECURRENCE_RELATION"(STATE)
        memo[STATE] = ans
        return ans

    memo = {}
    return dp("STATE_FOR_WHOLE_INPUT")

File: javascript/15_dynamic_programming.js
// Dynamic programming: top-down memoization

let fn = arr => {
    let dp = STATE => {
        if (BASE_CASE) {
            return 0;
        }

        if (memo[STATE] != -1) {
            return memo[STATE];
        }

        let ans = RECURRENCE_RELATION(STATE);
        memo[STATE] = ans;
        return ans;
    }

    let memo = ARRAY_SIZED_ACCORDING_TO_STATE;
    return dp(STATE_FOR_WHOLE_INPUT);
}